front_end/models/react_native/FuseboxBindingsModel.ts

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// Copyright 2024 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import * as SDK from '../../core/sdk/sdk.js';
+import type * as Common from '../../core/common/common.js';
+
+type JSONValue = null | string | number | boolean | {[key: string]: JSONValue} | JSONValue[];
+type DomainName = 'react-devtools';
+type DomainMessageListener = (message: JSONValue) => void;
+type BindingCalledEventTargetEvent = Common.EventTarget.EventTargetEvent<SDK.RuntimeModel.EventTypes[SDK.RuntimeModel.Events.BindingCalled]>;
+
+const RUNTIME_GLOBAL = '__FUSEBOX_REACT_DEVTOOLS_DISPATCHER__';
+
+export class FuseboxBindingsModel extends SDK.SDKModel.SDKModel {
+  private readonly domainToListeners: Map<DomainName, Set<DomainMessageListener>> = new Map();
+
+  constructor(target: SDK.Target.Target) {
+    super(target);
+
+    this.initialize(target);
+  }
+
+  private initialize(target: SDK.Target.Target): void {
+    const runtimeModel = target.model(SDK.RuntimeModel.RuntimeModel);
+    if (!runtimeModel) {
+      return;
+    }
+
+    runtimeModel.addEventListener(SDK.RuntimeModel.Events.BindingCalled, this.bindingCalled, this);
+    void this.enable(target);
+  }
+
+  private bindingCalled(event: BindingCalledEventTargetEvent): void {
+    const serializedMessage = event.data.payload;
+
+    try {
+      const {domain, message} = JSON.parse(serializedMessage);
+      this.dispatchMessageToDomainEventListeners(domain, message);
+    } catch (err) {
+      throw new Error('Failed to parse bindingCalled event payload:', err);
+    }
+  }
+
+  subscribeToDomainMessages(domainName: DomainName, listener: DomainMessageListener): void {
+    let listeners = this.domainToListeners.get(domainName);
+    if (!listeners) {
+        listeners = new Set();
+        this.domainToListeners.set(domainName, listeners);
+    }
+
+    listeners.add(listener);
+  }
+
+  unsubscribeFromDomainMessages(domainName: DomainName, listener: DomainMessageListener): void {
+    const listeners = this.domainToListeners.get(domainName);
+    if (!listeners) {
+      throw new Error(`Failed to unsubscribe from domain ${domainName}: no listeners were previously attached`);
+    }
```

Comment:
Is there a way for callers to check if there's an active subscription before calling this? (or we could do a warning instead?)

Generally I'd expect an unsubscribe function to be no-throw since the clean-up path may not always be symmetrical with the set-up path, e.g.

```
function onPanelOpen() {
  fuseboxBindingModel.subscribe();
}

function onPanelClosed() {
  fuseboxBindingModel.unsubscribe();
}

function onRuntimeDestory() {
  fuseboxBindingModel.unsubscribe();
}
```

We might also be preventing other clean-up code from running, e.g.

```
function cleanup() {
  fuseboxBindingModel.unsubscribe();  // crashed
  webSocket.close();  // not called
}
```

(In practice, it'll just be our own code listening so it's up to you)

Reply:
Makes sense, this will also make the method idempotent.